### ISSUE TITLE
Duplicate imports

### ISSUE THREAD

**Issue by krlmlr** [rticulate/import#9]
Currently, `import` happily allows importing the same symbol from more than one package, the last import wins:

```
> import::from("plyr", "arrange")
> environment(arrange)
<environment: namespace:plyr>
> import::from("dplyr", "arrange")
> environment(arrange)
<environment: namespace:dplyr>
> import::from("plyr", "arrange")
> environment(arrange)
<environment: namespace:plyr>
```

Would you consider a "safe mode" where this usage triggers an error?

**Comment by smbache** [comment]
let me give that some thought...

Also, what would be ideal here:

```R
import::from("plyr", "arrange", .into = "plyrstuff")
import::from("dplyr", "arrange", .into = "dplyrstuff")
```

Here one explicitely separates them, but still one is before the other in the search path. But one can do:

```R
p <- import::from("plyr", "arrange", .into = "plyrstuff")
d <- import::from("dplyr", "arrange", .into = "dplyrstuff")

# now
p$arrange
# and 
d$arrange
# refers to the correct versions.
```

But I'm thinking that it's at least worth enforcing only one import of the same name in each search path location...

**Comment by gaborcsardi** [comment]
First I thought the `.into` is overkill, and people can just rename. But on a second thought I quite like it. Especially if you can include a non-scalar character vector with the function names, i.e. 

```r
p <- import::from("plyr", c("arrange", "each", "empty"), .into = "plyrstuff")
```

But then maybe you would want to avoid putting `arrange`, etc. on the search path if `.into` is given. Maybe.

**Comment by krlmlr** [comment]
Maybe ;-) I was just thinking along the same lines.

How about

```
import::from("plyr", "arrange", .namespace = "p")
```

which would create a new environment (if missing) in the default `"imports"` namespace, and populate it? Then, we can still use `p$arrange` but the search path isn't polluted.

**Comment by smbache** [comment]
How is that better than 

```R
p <- import::from("plyr", "arrange", "each", "empty", .into = "plyrstuff")
```

**Comment by drknexus** [comment]
It does seem that it is as you said...
> at least worth enforcing only one import of the same name in each search path location
Is that something you are interested in working on?  If not, would you accept a PR if it happened to bubble up in my priorities?

**Comment by krlmlr** [comment]
The new ["strict" package](https://github.com/hadley/strict) provides an alternative to selective importing.

**Comment by drknexus** [comment]
@krlmlr interesting. That package has come a long way in a very short time! All other things being equal, I think my team would prefer imports with this python flavor (esp for conflict resolution). But, I'll float strict also and see where things go. Thanks for pointing out the new functionally!

**Comment by smbache** [comment]
@drknexus I have made imports a bit more strict - now if you try to import and reuse a name that exists at the destination, an error will be raised unless the environment of the existing and the new object is the same. The reason for not checking for the two being identical, is that this will allow to re-import an updated version of a function from the same environment, which often is useful when importing from a script module. Try it out and see if this is in line with what you want.

01c27bf3c7ffda2f3e81656fe5d4ca1d5e109d34

**Comment by mmuurr** [comment]
This [safe assignment](https://github.com/smbache/import/blob/master/R/safe_assign.R) just caused a minor problem with at least one pattern that I use.

In a top-level script, one might do something like this:
```
library(DBI)
library(myDBI)
```
Where `myDBI` exports some functions that are _purposefully_ named identically to a few that are in `DBI`. These might be `DBI` calls with a few of the defaults changed, for example. Rather than re-exporting _all_ of `DBI`'s exported objects, only those that are being 'masked' are included in `myDBI`.

When trying to replicate this pattern using `import`'s 'module' functionality, I was hoping to do something like this:
```
import::from(getNamespaceExports("DBI"), .from = "DBI", .character_only = "true")  ## import all of DBI
import::from(getNamespaceExports("myDBI"), .from = "myDBI", .character_only = "true")  ## import all of myDBI
```
When the masking functions are then encountered in `myDBI`, the 'safe assignment' fails, where I'd _prefer_ (in this use case) to do the overriding that sparked this issue's thread in the first place :-)

One way around this, obviously, would be for `myDBI` to fully re-export all of `DBI` and just removed `DBI` from the pattern altogether. I tend to not like doing this, because I want people (in my org) to generally use standard packages and then use some internal packages as 'enhancements' when appropriate.

Another way around this might be to make the 'strict' assignment _optionally_ strict.
I'd be happy to spin up a branch and submit a PR with that option (defaulting to the string-mode), if you support that idea, @smbache.

**Comment by smbache** [comment]
@mmuurr i think strict is fine and shows that there is some dangerous aspect to your code. IMO your identically named functions could confuse readers that expect different behaviour. 

That said, you can use two different destinations with `.into` and the search path will determine which masks which, should solve your problem.

**Comment by mmuurr** [comment]
@smbache The `.into` arg was my first thought, but it appears it's not used as such when in 'module' (i.e. 'script') mode?
For example, here's `my-module.R`:
```R
import::from("DBI", getNamespaceExports("DBI"), .into = "foo_DBI", .character_only = TRUE)
import::from("myDBI", getNamespaceExports("myDBI"), .into = "foo_myDBI", .character_only = TRUE)

f <- function() NULL
```
And then when trying to import from that module:
```R
> import::from("my-module.R", "f")

 Error in as.environment(where) :
   no item called "foo_DBI" on the search list
```

**Comment by mmuurr** [comment]
Thinking about this a bit more, I guess I hadn't realized `import` was bring all imported objects into a single environment (when not explicitly specifying the `.into` argument). I'd assumed (naïvely) that successive imports from distinct scripts/packages were 'splicing' new environments into the environment chain, like `attach`/`library`/etc. (and the `modules` package) do, and thus 'masking' duplicates via precedence in the environment chain.

I think there are some scenarios where that masking is useful ... e.g. alternate implementations (that might take advantage of hardware accelerations only available on some machines) that are drop-in replacements returning identical values but through different code-paths. I completely agree with you though that actually 'overwriting' the objects in the same namespace is a bad idea, and armed with a better understanding of how `import` creates environments I'll rethink my module pattern (and possibly use [`modules`](https://cran.r-project.org/web/packages/modules/index.html) for the 'script' use case).

Thanks much for the discussion & help!

**Comment by smbache** [comment]
I’d say avoid dependencies unless it’s necessary.
Is the bug mentioned related to [this issue](https://github.com/smbache/import/issues/28)?

**Comment by mmuurr** [comment]
> Is the bug mentioned related to this issue?

@smbache, I don't think so. I'm running on the branch that fixes that specific issue, but within modules (i.e. when `is_script` is `TRUE`), the bug mentioned above is still occurring.

**Comment by smbache** [comment]
In any case, the behaviour feels like a bug

**Comment by torfason** [comment]
It looks to me like there although `import` is smart about not allowing duplicate imports into the same namespace, it is also smart and allows imports to *different* namespaces.  This is the way things work in master (available using `remotes::install_github("smbache/import")`:

``` r
# Fine to import same object into different namespaces
import::from(plyr, mutate, .into="import:plyr")
import::from(dplyr, mutate, .into="import:dplyr")

# Last one wins
mutate
#> function (.data, ...) 
#> {
#>     UseMethod("mutate")
#> }
#> <bytecode: 0x7f802e872158>
#> <environment: namespace:dplyr>

# But don't try to overwrite new object into same namespace
import::from(dplyr, mutate, .into="import:plyr")
#> Error: Cannot assign name to different value in the given environment. Name already in use.
```

This feels like the right behavior, and I do not think we should change it further. 

Any remaining issues are probably caused by the recursive use of import (using `import::from()` in a module, that is then itself imported using `import::from()`), that issue can be fixed by using `import::here()` in the module file insted of `import::from()`. Any further discussions about the recursive import issue should probably be directed to #13 for now.

**Comment by torfason** [comment]
This issue, as well as other issues relating to recursive `import::from()` (such as #13 and #34) have now been fixed, the fix is reflected in the current release, 1.2.0, available on CRAN.
